sql/core/src/main/scala/org/apache/spark/sql/execution/joins/OuterJoin.scala

```diff
@@ -111,52 +122,43 @@ trait HashOuterJoin {
   // iterator for performance purpose.
 
   protected[this] def leftOuterIterator(
-      key: InternalRow,
       joinedRow: JoinedRow,
       rightIter: Iterable[InternalRow],
       resultProjection: InternalRow => InternalRow): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
-      if (!key.anyNull) {
-        val temp = if (rightIter != null) {
-          rightIter.collect {
-            case r if boundCondition(joinedRow.withRight(r)) => resultProjection(joinedRow).copy()
```

Comment:
This problem existed in the old code, but I should fix it here: I don't think that we need to perform this `copy()` here.

Also, I don't see why we need to construct `val ret: Iterable[InternalRow]` here only to turn around and immediately call `ret.iterator` on it.  This seems like it will cause a lot of inefficient copying and materialization.